File: src/client/java/fr/hardel/asset_editor/client/javafx/components/page/recipe/RecipeTreeBuilder.java
package fr.hardel.asset_editor.client.javafx.components.page.recipe;

import fr.hardel.asset_editor.client.javafx.components.ui.tree.TreeNodeModel;
import fr.hardel.asset_editor.client.javafx.lib.data.RecipeTreeData;
import net.minecraft.resources.Identifier;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class RecipeTreeBuilder {

    public static TreeNodeModel build(List<RecipeEntry> elements) {
        TreeNodeModel root = new TreeNodeModel();
        root.setCount(elements.size());

        for (RecipeTreeData.RecipeBlockConfig block : RecipeTreeData.RECIPE_BLOCKS) {
            if (block.special()) continue;

            List<String> typeStrings = block.recipeTypes().stream().map(Identifier::toString).toList();
            ArrayList<RecipeEntry> matching = new ArrayList<>();
            for (RecipeEntry element : elements) {
                if (typeStrings.contains(element.type())) {
                    matching.add(element);
                }
            }

            boolean hasSubTypes = block.recipeTypes().size() > 1;
            TreeNodeModel blockNode = new TreeNodeModel();
            blockNode.setCount(matching.size());

            if (hasSubTypes) {
                for (Identifier recipeType : block.recipeTypes()) {
                    String typeStr = recipeType.toString();
                    ArrayList<RecipeEntry> subMatching = new ArrayList<>();
                    for (RecipeEntry element : matching) {
                        if (typeStr.equals(element.type())) {
                            subMatching.add(element);
                        }
                    }
                    if (subMatching.isEmpty()) continue;
                    TreeNodeModel subNode = new TreeNodeModel();
                    subNode.setCount(subMatching.size());
                    for (RecipeEntry element : subMatching) {
                        subNode.identifiers().add(element.uniqueId());
                    }
                    blockNode.children().put(typeStr, subNode);
                }
            } else {
                for (RecipeEntry element : matching) {
                    blockNode.identifiers().add(element.uniqueId());
                }
            }

            root.children().put(block.blockId().toString(), blockNode);
        }

        return root;
    }

    public static Map<String, Identifier> folderIcons() {
        LinkedHashMap<String, Identifier> icons = new LinkedHashMap<>();
        for (RecipeTreeData.RecipeBlockConfig block : RecipeTreeData.RECIPE_BLOCKS) {
            if (block.special()) continue;
            icons.put(block.blockId().toString(), block.icon());
            for (Identifier type : block.recipeTypes()) {
                icons.put(type.toString(), block.icon());
            }
        }
        return icons;
    }

    public record RecipeEntry(String uniqueId, String type) {
    }

    private RecipeTreeBuilder() {
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/page/recipe/RecipeTreeBuilder.kt
package fr.hardel.asset_editor.client.compose.components.page.recipe

import fr.hardel.asset_editor.client.compose.components.ui.tree.TreeNodeModel
import fr.hardel.asset_editor.client.compose.lib.data.RecipeTreeData
import net.minecraft.client.resources.language.I18n
import net.minecraft.resources.Identifier

object RecipeTreeBuilder {

    fun build(elements: List<RecipeEntry>): TreeNodeModel {
        val root = TreeNodeModel()
        root.count = elements.size

        RecipeTreeData.RECIPE_ENTRIES.forEach { entry ->
            if (entry.special) {
                return@forEach
            }

            val typeStrings = entry.recipeTypes.map(Identifier::toString)
            val matching = elements.filter { element -> typeStrings.contains(element.type) }
            val entryNode = TreeNodeModel()
            entryNode.count = matching.size
            entryNode.label = I18n.get(entry.translationKey)

            if (entry.recipeTypes.size > 1) {
                entry.recipeTypes.forEach { recipeType ->
                    val type = recipeType.toString()
                    val subMatching = matching.filter { element -> element.type == type }
                    if (subMatching.isEmpty()) {
                        return@forEach
                    }
                    val subNode = TreeNodeModel()
                    subNode.count = subMatching.size
                    subNode.label = I18n.get("recipe:crafting.${recipeType.path}.name")
                    subMatching.forEach { element -> subNode.identifiers += element.uniqueId }
                    entryNode.children[type] = subNode
                }
            } else {
                matching.forEach { element -> entryNode.identifiers += element.uniqueId }
            }

            root.children[entry.entryId.toString()] = entryNode
        }

        return root
    }

    fun folderIcons(): Map<String, Identifier> =
        LinkedHashMap<String, Identifier>().apply {
            RecipeTreeData.RECIPE_ENTRIES.forEach { entry ->
                if (entry.special) {
                    return@forEach
                }
                put(entry.entryId.toString(), entry.folderIcon())
                entry.recipeTypes.forEach { type -> put(type.toString(), entry.folderIcon()) }
            }
        }

    data class RecipeEntry(val uniqueId: String, val type: String)
}
